metricbeat/helper/server/tcp/tcp.go

```diff
@@ -96,22 +98,46 @@ func (g *TcpServer) watchMetrics() {
 			logp.Err("Unable to accept connection due to error: %v", err)
 			continue
 		}
-		defer func() {
-			if conn != nil {
-				conn.Close()
-			}
-		}()
 
-		length, err := conn.Read(buffer)
+		go g.handle(conn)
+	}
+}
+
+func (g *TcpServer) handle(conn net.Conn) {
+	logp.Debug("tcp", "Handling new connection...")
+
+	// Close connection when this function ends
+	defer func() {
+		if conn != nil {
+			conn.Close()
+		}
+	}()
+
+	// Get a new reader with buffer size as the same as receiveBufferSize
+	bufReader := bufio.NewReaderSize(conn, g.receiveBufferSize)
+
+	for {
+		// Read tokens delimited by delimiter
+		bytes, err := bufReader.ReadBytes(g.delimiter)
```

Comment:
I think this is what is blocking tests, we should support also reading single tokens that don't include a delimiter.

Reply:
@jsoriano i have fixed the faulty test. I would argue if we should accept tokens that dont have a delimiter.

Reply:
I'm afraid that we can break something, I guess that currently we accept single graphite metrics sent without a delimiter at the end...
We should at least mention it in the breaking changes section.

Reply:
@vjsamuel I guess that the thing here is that if someone sends something without the delimiter at the end, `ReadBytes()` is going to return `EOF` as error, could we just consider it also valid `if len(bytes) > 0 && err == i.EOF`?

Reply:
@jsoriano the spec requires a delimiter and till date we were accepting data that had a `\n` by default but not accepting multiple entries that were separated by `\n` on the same payload.
https://graphite.readthedocs.io/en/latest/feeding-carbon.html#step-3-understanding-the-graphite-message-format

Reply:
so it wouldnt constitute a breaking change

Reply:
You are right, I have also tested it with graphite and it doesn't accept metrics without a trailing newline.

Reply:
thanks :)